jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java

```diff
@@ -0,0 +1,395 @@
+//
+
+package org.eclipse.jetty.util;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+import org.eclipse.jetty.util.log.Log;
+import org.eclipse.jetty.util.log.Logger;
+
+/**
+ * A fast container of poolable objects, with optional support for
+ * multiplexing and thread-local caching.
+ * @param <T>
+ */
+public class Pool<T> implements AutoCloseable
+{
+    private static final Logger LOGGER = Log.getLogger(Pool.class);
+
+    private final List<Entry> sharedList = new CopyOnWriteArrayList<>();
+    private final Collection<List<Entry>> allCaches = new CopyOnWriteArrayList<>();
+    private final ThreadLocal<List<Entry>> cache;
+    private final Lock closeLock = new ReentrantLock();
+    private final int cacheSize;
+    private volatile boolean closed;
+    private volatile int maxMultiplex = 1;
+    private volatile int maxUsageCount = -1;
+
+    /**
+     * Construct a Pool with the specified thread-local cache size.
+     *
+     * @param cacheSize the thread-local cache size. A value of 0 means the cache is disabled.
+     */
+    public Pool(int cacheSize)
+    {
+        if (cacheSize < 0)
+            throw new IllegalArgumentException("Cache size must be >= 1");
+        this.cacheSize = cacheSize;
+        if (cacheSize > 0)
+        {
+            this.cache = ThreadLocal.withInitial(() ->
+            {
+                List<Entry> list = new ArrayList<>(cacheSize);
+                allCaches.add(list);
+                return list;
+            });
+        }
+        else
+        {
+            this.cache = null;
+        }
+    }
+
+    public int getMaxMultiplex()
+    {
+        return maxMultiplex;
+    }
+
+    public final void setMaxMultiplex(int maxMultiplex)
+    {
+        if (maxMultiplex < 1)
+            throw new IllegalArgumentException("Max multiplex must be >= 1");
+        this.maxMultiplex = maxMultiplex;
+    }
+
+    public int getMaxUsageCount()
+    {
+        return maxUsageCount;
+    }
+
+    public final void setMaxUsageCount(int maxUsageCount)
+    {
+        if (maxUsageCount == 0)
+            throw new IllegalArgumentException("Max usage count must be != 0");
+        this.maxUsageCount = maxUsageCount;
+    }
+
+    public Entry newEntry(T pooled)
+    {
+        return new Entry(pooled);
+    }
+
+    /**
+     * Add a new object to the pool for others to acquire.
+     *
+     * @param entry an object to add to the pool
+     */
+    public void add(Entry entry)
+    {
+        closeLock.lock();
+        try
+        {
+            if (closed)
+                throw new IllegalStateException("Pool has been closed");
```

Comment:
I'd return `false` here. Cases where the client is shutting down when a connection is being added are not uncommon.